File: code_snippets/ts/src/use_cases/microservices/subscription_service.ts
import * as restate from "@restatedev/restate-sdk";

// <start_here>
// <mark_1>
const subscriptionService = restate.service({
  // </mark_1>
  name: "SubscriptionService",
  handlers: {
    // <mark_1>
    add: async (ctx: restate.Context, req: SubscriptionRequest) => {
      // </mark_1>
      const paymentId = ctx.rand.uuidv4();

      // <mark_2>
      const payRef = await ctx.run(() =>
        createRecurringPayment(req.creditCard, paymentId)
      );
      // </mark_2>

      for (const subscription of req.subscriptions) {
        // <mark_2>
        await ctx.run(() =>
          createSubscription(req.userId, subscription, payRef)
        );
        // </mark_2>
      }
    },
  },
});

export type SubscriptionService = typeof subscriptionService;

restate.endpoint().bind(subscriptionService).listen(9080);
// <end_here>

type SubscriptionRequest = {
  creditCard: string;
  userId: any;
  subscriptions: any[];
};

function createRecurringPayment(creditCard: string, paymentId: string) {}

function createSubscription(userId: any, subscription: any, payRef: void) {
  return undefined;
}

File: code_snippets/python/src/use_cases/microservices/subscription_service.py
import uuid

import restate
from pydantic import BaseModel
from restate import Service, Context


class SubscriptionRequest(BaseModel):
    user_id: str
    credit_card: str
    subscriptions: list[str]


# <start_here>
# <mark_1>
subscription_service = Service("SubscriptionService")
# </mark_1>


# <mark_1>
@subscription_service.handler()
async def add(ctx: Context, req: SubscriptionRequest):
    # </mark_1>
    payment_id = await ctx.run("payment id", lambda: str(uuid.uuid4()))

    # <mark_2>
    pay_ref = await ctx.run(
        "recurring payment",
        lambda: create_recurring_payment(req.credit_card, payment_id),
    )
    # </mark_2>

    for subscription in req.subscriptions:
        # <mark_2>
        await ctx.run(
            "subscription",
            lambda: create_subscription(req.user_id, subscription, pay_ref),
        )
        # </mark_2>


app = restate.app([subscription_service])
# <end_here>


def create_recurring_payment(credit_card, payment_id):
    return None


def create_subscription(user_id, subscription, pay_ref):
    return None
